Scripts/Python/plasma/generate_stubs.py

```diff
@@ -0,0 +1,382 @@
+    yield f'"""{__doc__}"""'
+
+    yield ""
+    yield "# NOTE: This stub file was generated automatically from Plasma's Python interface."
+    yield "# Do not edit this file manually."
+    yield "# To change any of the docstrings or function signatures,"
+    yield "# edit the corresponding C++ glue code in pfPython."
+    yield "# If the Python interface has changed, regenerate these stubs"
+    yield "# by running the following call in the in-game Python console:"
+    yield '# >>> __import__("generate_stubs").run()'
+    yield ""
+
+    # Now the actual module docstring (if one exists).
+    # Temporarily disabled, because the docstring is currently occupied by the license header,
+    # so a string literal placed here would be considered a regular statement and not a docstring.
+    # This is also important for the __future__ import,
+    # which loses its effect if preceded by any statement other than a docstring.
+    if False and getattr(module, "__doc__", None) is not None:
+        yield f'"""{module.__doc__}"""'
+        yield ""
+
+    # Hardcoded imports for type annotations:
+    yield "from __future__ import annotations"
+    if module.__name__ == "Plasma":
+        # Some parameter default values in the Plasma module
+        # use constant values from PlasmaConstants.
+        # Not sure how to implement a good generic solution to this problem,
+        # so for now, just hardcode this dependency... :(
+        yield "from PlasmaConstants import *"
+    yield "from typing import *"
+
+    for name, value in iter_attributes(module):
+        if name.startswith("__"):
+            continue
+
+        yield ""
+        if isinstance(value, PlasmaConstants.Enum):
+            yield from generate_enum_stub(name, value)
+        elif isinstance(value, type):
+            yield from generate_class_stub(name, value)
+        elif callable(value):
+            signature, doc = parse_type_from_doc(value.__doc__)
+            yield from generate_function_stub("function", name, signature, doc)
+        else:
+            yield f"{name}: {format_qualified_name(type(value), module.__name__)} = ... # = {value!r}"
+
+def run(dest_dir: str = "plasma_stubs_generated") -> None:
+    try:
+        os.mkdir(dest_dir)
+    except FileExistsError:
+        pass
+
+    # Note: Keep this list in sync with the IInitBuiltinModule calls
+    # in PythonInterface::initPython in pfPython/cyPythonInterface.cpp!
+    for module in all_plasma_modules:
+        # Generate all the lines first and only write the file afterwards
+        # to avoid leaving behind a half-written file if generate_module_stub throws an exception.
+        lines = list(generate_module_stub(module))
+        stub_file = os.path.join(dest_dir, module.__name__ + ".py")
+        with open(stub_file, "w", encoding="utf-8", newline="\n") as f:
+            for line in lines:
+                f.write(line + "\n")
```

Comment:
Using the `pathlib` module would make all of this quite a bit less tedious, and is highly preferable IMO to all this manual path string fiddling.

Reply:
Unfortunately, this module/script needs to be run from inside the game, and `pathlib` cannot be used there, because it unconditionally imports `urllib`:

```python-traceback
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File ".../python/system\pathlib.py", line 14, in <module>
    from urllib.parse import quote_from_bytes as urlquote_from_bytes
ModuleNotFoundError: No module named 'urllib'
```